050-unidirectional-topic-operator.md

```diff
@@ -0,0 +1,680 @@
+>     ```
+> 
+> 2. The user then deletes the resource (e.g. `kubectl delete kafkatopic foo`).
+>     Because of the presence of a `metadata.finalizer` the resource remains present.
+>     Instead Kube adds the `metadata.deletionTimestamp` field.
+> 
+>     ```yaml=
+>     metadata:
+>       generation: 124
+>       name: foo
+>       finalizer: 
+>       - strimzi.io/topic-operator
+>       deletionTimestamp: 20230301T000000.000
+>     spec:
+>       # ...
+>     status:
+>       observedGeneration: 124 # == generation => it is not syncrhonized
+>     ```
+> 
+> 3. The operator notices the `metadata.deletionTimestamp` field.
+>    1. If there is more than one `KafkaTopic` which is trying to manage this topic then deletion from Kafka is not attempted;
+>    the operator will remove the finalizer and Kube will remove the resource; upon timed reconciliation the condition on the other `KafkaTopic` will be removed and reconciliation via that other `KafkaTopic` will proceed as normal.
+>    2. If deletion is disallowed via `delete.topic.enable=false` broker configuration then the operator will remove the finalizer and Kube will remove the resource. The topic in Kafka will thus become unmanaged.
+>    3. Otherwise, the operator attempted to deletes the topic from Kafka.
+>        1. If deletion succeeds the operator will remove `strimzi.io/topic-operator` from the `metadata.finalizer`.
+>        2. If deletion failed with `UNKNOWN_TOPIC_OR_PARTITION` (i.e. the topic didn't exist) the operator will continue to remove the finalize. I.e. this case is not treated as an error.
+>        3. If deletion failed with any other error this is reported via a condition in the `status.conditions` and the finalizer is not removed.
+---
+
+Similarly to how it is possible to "manage" an existing unmanaged topic in Kafka, we will also provide a mechanism for "unmanaging" a managed topic, via a `spec.managed` flag.
+This will allow a `KafkaTopic` to be deleted **without** the topic in Kafka being deleted.
+This can be useful operationally, for example to change the `metadata.name` of a `KafkaTopic`.
+
+> An alternative to `spec.managed` would be "explicit deletion" via a `spec.delete` flag.
+> This alternative was rejected because it's incompatible with gitops.
+
+---
+
+#### Example 4: `KafkaTopic` deletion WITHOUT propagation to Kafka
+
+> 1. Assume the topic is currently Kube-managed
+> 
+>     ```yaml=
+>     metadata:
+>       generation: 123
+>       name: foo
+>       finalizer: 
+>       - strimzi.io/topic-operator
+>     spec:
+>       # ...
+>     status:
+>       observedGeneration: 123 # == generation => it is Kube-managed
+>     ```
+> 
+> 2. The user changes the new `spec.managed` field to `False` so that the topic is no longer Kube-managed.
+> 
+>     ```yaml=
+>     metadata:
+>       generation: 124
+>       name: foo
+>       finalizer: 
+>       - strimzi.io/topic-operator
+>     spec:
+>       managed: False # do not synchronize to Kafka
```

Comment:
The user operator uses a `strimzi.io/pause-reconciliation` annotation to do a similar thing. We should use that (if the semantics are the same), or be explicit about why we've not using it (if the semantics of `spec.managed` are necessarily different)

Reply:
This annotation is used in Cluster Operator as well. When set, the condition in `.status` is set to `Paused`.

Reply:
What happens if you delete a paused CR?

Reply:
I actually don't know for sure about User Operator. I expect the deletion will proceed since eventually in the next timer based reconciliation, it will not remember anymore that this was paused user. For the resources such as `Kafka`, the Kube GC will wipe the resources right away.

Reply:
Since the UO doesn't use finalizers if a paused `KafkaUser` is deleted then the user will be removed from Kafka. So the behaviour seems pretty similar. Maybe we should use the annotation instead to be consistent. That would also fit nicely with using the Paused annotation instead of removing the `spec.topicName`. WDYT?

Reply:
You mean instead of having the `.spec.managed` field to have some annotation `strimzi.io/managed: false`? That works for me, but not sure it makes a big difference. 

Or do you want to use the paused-reconciliation annotation for this but not delete the topic from Kafka? TBH, it would work for me as well. But it would be a different behavior from the User or Cluster operator as there will the user or cluster be really deleted.

The TO obviously works differently as it will not have any _reconcile all topics_ method as it doe snot manage all topics. In the UO, it is not the deletion of the paused KafkaUser which triggers the deletion. It is the periodical reconciliation and having the CRs as a single source of truth. Similarly, It is not the CO deleting the Kafka cluster which is paused - it is Kubernetes GC. So code-wise, the TO behavior would be the same. But from the user perspective it would differ.

Reply:
On balance: 

* I think the difference between paused and unmanaged is fundamental and important, so let's not use `strimzi.io/pause-reconciliation`.
* But it's probably more consistent to use an annotation for managed than to put it in the `spec`. So I'll change it to `strimzi.io/managed` with a boolean value, defaulting to true.
* I thought too about the `Paused` condition and whether to use that or a `Managed=false` condition. But the way the operator works it will simply apply the different logic on the next reconciliation: It's not a piece of state that needs to be reconciled. i.e. it's just not possible for a resource with the annotation to behave transiently in a way that's incompatible with the annotation.